### DISCUSSION TITLE
go modules: add option to ignore replace directives

### DISCUSSION THREAD

**Discussion by paralin** [renovatebot/renovate#20917] [Suggest an Idea]
### Type of discussion.

I'm proposing an idea

### Tell us more.

Currently renovate sometimes updates my replaced dependencies to latest:

```
replace (
	github.com/sirupsen/logrus => github.com/aperturerobotics/logrus v1.9.1-0.20221224130652-ff61cbb763af // myfork
)
```

is replaced with...

```
replace (
	github.com/sirupsen/logrus => github.com/aperturerobotics/logrus v1.9.2 // myfork
)
```

### Proposed solution

Please add an option to ignore the replace() directives.

Reproduce: https://github.com/paralin/renovate-go-replace-statements/pull/2

### Reading from comments

The below is an additional somewhat obscure feature proposal and probably will not be implemented:

I want renovate to read the comment on the same line, if it's not "indirect" (which go adds automatically to some deps), and if the comment is a single word, I want to use the contents of the comment as the version:

```
replace (
	github.com/sirupsen/logrus => github.com/aperturerobotics/logrus myfor // myfork
)
```

And then run "go mod tidy" which resolves to a specific ref.

**Comment by rarkins** [comment]
I'm not a regular go user but this sounds a bit hacky/risky. You may not be the only person in the world who adds comments to a `go.mod` but now you are asking Renovate to imply meaning from them which may not intended by everyone else.

If you are not updating your fork often, you'd be better off disabling it from updates for Renovate and instead doing it manually.

**Reply by paralin** [reply]
I'm not the only one to use comments this way. The go tool actually uses them to mark indirect dependencies as well.

It would be great to just add an option similar to "massageGoMod" which enables this behavior.

**Reply by rarkins** [reply]
Unfortunately, if we added config options any time it benefited one person, Renovate's config would become too polluted. Naturally the way _the tool_ does things is most important, but equating yourself to literally `go` itself is an unfair comparison.

I assume that a quick workaround for this is to disable that dependency, e.g. using `ignoreDeps`. If you find it easier to switch tools than add a line of config, so be it.

**Reply by paralin** [reply]
> Naturally the way the tool does things is most important, but equating yourself to literally go itself is an unfair comparison.

I'm just saying that clobbering replace statements is very inconvenient and probably contrary to a lot of projects workflows. The replace statement is intended to replace with a specific version, not "latest".

After all, this is the default behavior for dependabot

**Comment by rarkins** [comment]
Let's take a step back to make sure I understand fully. At which times is your replace statement being "updated"?
- Renovate does this explicitly/intentionally, it thinks you want 1.9.2, or
- This change happens - seemingly unintentionally by Renovate - as part of other upgrades

**Reply by paralin** [reply]
Renovate is creating pull requests which change the version of the modules in the replace statements to the "latest" version (this is a keyword in go for the most recent release). If I enable batch pull requests these changes are also bundled into the batch pr as well. Example: https://github.com/aperturerobotics/controllerbus/commit/d302b24f0a2f65d1d33b19722444ef18ba809acd

Theoretically I could list all of the dependencies I've replaced in the ignore statements but that's a level of micromanagement that is too much for my use case, I'd rather use dependabot than have to keep the replace statements and the ignore list synced across all my repos at risk of merging a update that breaks a replace statement as part of another batch without realizing.

By the way, thanks for creating renovate. It's a very cool bot and the docs / noise reduction features definitely have an edge on dependabot.

**Reply by rarkins** [reply]
Please directly answer my previous question so there's no confusion

**Reply by paralin** [reply]
Renovate is creating pull requests which change the version of the modules in the replace statements to the "latest" version.

Renovate does this explicitly/intentionally with a PR for that dependency.

This is not what I want, I'd rather have the replace statements be untouched.

**Reply by rarkins** [reply]
That should be easy enough to provide a solution for. Can you create a minimal reproduction of it? I assume it's as simple as a repo with a `go.mod` (no `go.sum` or go code even necessary) with a replacement dependency, which results in a mistaken upgrade

**Reply by paralin** [reply]
@rarkins Reproduced the issue: https://github.com/paralin/renovate-go-replace-statements/pull/2 - that repository is a simple reproduction which has 1 replace statement which should be ignored rather than updated. Thanks for having a look.

**Reply by rarkins** [reply]
Please merge https://github.com/paralin/renovate-go-replace-statements/pull/4 and confirm if it fixes the problem

**Reply by paralin** [reply]
@rarkins That worked! Thanks so much for your quick response! I will apply this to my other projects and switch back to renovate.
